File: samples/java/DeliveryDispatch/Server/CourierSpotNode/src/main/java/systems/zlink/samples/deliverydispatch/server/courierspotnode/spots/handlers/OfferDeliveryActorHandler.java
package systems.zlink.samples.deliverydispatch.server.courierspotnode.spots.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkEntrySpotActorSendHandler;
import systems.zlink.samples.deliverydispatch.server.courierspotnode.CourierActor;
import systems.zlink.samples.deliverydispatch.server.courierspotnode.spots.CourierEntrySpot;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

/**
 * The offer, in the courier actor's own turn: push it to the bound session and return (common
 * sample spec section 7.4).
 */
public final class OfferDeliveryActorHandler
        implements ZLinkEntrySpotActorSendHandler<
                CourierEntrySpot, CourierActor, Messages.OfferDeliveryMsg> {
    @Override
    public CompletionStage<Void> handle(
            CourierEntrySpot entrySpot,
            CourierActor actor,
            ZLinkMessageContext context,
            Messages.OfferDeliveryMsg message) {
        return actor.offer(message)
                .thenRun(
                        () ->
                                System.out.println(
                                        "deliverydispatch courier-actor: offer pushed delivery="
                                                + message.deliveryId()
                                                + " courier="
                                                + actor.actorId()
                                                + " attempt="
                                                + message.attempt()));
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CourierSpotNode/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/courierspotnode/spots/handlers/OfferDeliveryActorHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.spots.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorSendHandler
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.CourierActor
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.spots.CourierEntrySpot
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.OfferDeliveryMsg

/**
 * The offer, in the courier actor's own turn: push it to the bound session and return (common
 * sample spec section 7.4).
 */
class OfferDeliveryActorHandler :
    ZLinkSuspendingEntrySpotActorSendHandler<CourierEntrySpot, CourierActor, OfferDeliveryMsg> {
    override suspend fun handle(
        entrySpot: CourierEntrySpot,
        actor: CourierActor,
        context: ZLinkMessageContext,
        message: OfferDeliveryMsg,
    ) {
        actor.offer(message)
        println(
            "deliverydispatch courier-actor: offer pushed delivery=${message.deliveryId} " +
                "courier=${actor.actorId()} attempt=${message.attempt}"
        )
    }
}
